test/videoframeprocessor/DefaultVideoTransformDevice.test.ts

```diff
@@ -0,0 +1,467 @@
+        }
+        destroy(): Promise<void> {
+          return;
+        }
+        process(_buffers: VideoFrameBuffer[]): Promise<VideoFrameBuffer[]> {
+          throw new Error('Method not implemented.');
+        }
+      }
+      let called = true;
+      let notCalled = true;
+      class Observer implements DefaultVideoTransformDeviceObserver {
+        processingDidFailToStart(): void {
+          called = true;
+        }
+      }
+
+      class Observer2 implements DefaultVideoTransformDeviceObserver {
+        processingDidStart(): void {
+          notCalled = false;
+        }
+      }
+      const obs = new Observer();
+      const processor = new WrongProcessor();
+      const device = new DefaultVideoTransformDevice(logger, 'test', [processor]);
+      device.addObserver(obs);
+      device.addObserver(new Observer2());
+      device.applyProcessors(mockVideoStream);
+
+      new TimeoutScheduler(300).start(() => {
+        expect(called).to.equal(true);
+        expect(notCalled).to.equal(true);
+        device.stop();
+      });
+
+      new TimeoutScheduler(350).start(() => {
+        done();
+      });
+    });
+
+    it('processingDidFailToStart', done => {
+      class WrongProcessor implements VideoFrameProcessor {
+        name(): string {
+          return '';
+        }
+        destroy(): Promise<void> {
+          return;
+        }
+        process(_buffers: VideoFrameBuffer[]): Promise<VideoFrameBuffer[]> {
+          throw new Error('Method not implemented.');
+        }
+      }
+      const obs = new MockObserver();
+      const processor = new WrongProcessor();
+      const device = new DefaultVideoTransformDevice(logger, 'test', [processor]);
+      device.addObserver(obs);
+      device.applyProcessors(mockVideoStream);
+
+      new TimeoutScheduler(300).start(() => {
+        expect(obs.processingDidFailToStart.called).to.equal(true);
+        device.stop();
+      });
+
+      new TimeoutScheduler(350).start(() => {
+        device.removeObserver(obs);
+        done();
+      });
+    });
+
+    it('processingLatencyTooHigh', done => {
+      class Observer2 implements DefaultVideoTransformDeviceObserver {}
+      const obs = new MockObserver();
+
+      class WrongProcessor extends NoOpVideoFrameProcessor {
+        async process(buffers: VideoFrameBuffer[]): Promise<VideoFrameBuffer[]> {
+          await new Promise(resolve => setTimeout(resolve, (1000 / 15) * 3));
+          return buffers;
+        }
+      }
+
+      const procs = [new WrongProcessor()];
+      const device = new DefaultVideoTransformDevice(logger, 'test', procs);
+      device.addObserver(obs);
+      device.addObserver(new Observer2());
+      device.applyProcessors(mockVideoStream);
+      // simulate high latency, has to wait long enough for the callback to be fired.
+      new TimeoutScheduler(800).start(() => {
+        expect(obs.processingLatencyTooHigh.called).to.equal(true);
+        device.stop();
```

Comment:
… but note that you could just wait on the observer instead of setting a timer here. The whole test runs in a timeout.

Reply:
how do I await on observer ? 

something like 
```
      await new Promise<void>(async(resolve) => {
        class Observer implements VideoFrameProcessorPipelineObserver {
          processingDidStart(): void {
            resolve();
          }
        }
        const obs = new Observer();
        pipe.addObserver(obs);
        await pipe.setInputMediaStream(mockVideoStream);
      });

``` 
?

Reply:
how do I await on two callbacks ?

Reply:
For example (untested):

```typescript
class MyObserver extends Whatever {
  onFoo: sinon.stub(),
  onBar: sinon.stub(),
}

const observer = new MyObserver();

// Abstracting this into helpers is trivial.
const firstObserverCalled = new Promise((resolve, reject) => {
  observer.onFoo.callsFake(resolve);
});
const secondObserverCalled = new Promise((resolve, reject) => {
  observer.onBar.callsFake(resolve);
});

// If you want to impose a timeout, you can add Promise.race here.
const bothCalled = Promise.all([firstObserverCalled, secondObserverCalled]);

// Now either…
bothCalled.then(done);

// or, for an async test function…
await bothCalled;
```

Reply:
```typescript
function timeoutPromise(timeout: number): Promise<void> {
  return new Promise((_resolve, reject) => {
    setTimeout(timeout, reject);
  };
}

function called(stub: SinonStub, timeout: number = 1_000): Promise<void> {
  return Promise.race([
    new Promise((resolve, _reject) => { stub.callsFake(resolve); }),
    timeoutPromise(timeout),
  ]);
}

// Returns a promise that resolves if both observers are called within one second,
// and rejects if either takes longer than a second.
const bothCalled = Promise.all([called(observer.onFoo), called(observer.onBar)]);
```

Reply:
coolio, works like a charm, now I can remove all the timeout